app/src/main/java/org/mozilla/fenix/crashes/CrashContentIntegration.kt

```diff
@@ -0,0 +1,74 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.crashes
+
+import android.view.ViewGroup.MarginLayoutParams
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.mapNotNull
+import mozilla.components.browser.state.selector.findTabOrCustomTabOrSelectedTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.EngineState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.browser.toolbar.BrowserToolbar
+import mozilla.components.lib.state.helpers.AbstractBinding
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+import org.mozilla.fenix.components.AppStore
+import org.mozilla.fenix.components.Components
+import org.mozilla.fenix.utils.Settings
+
+/**
+ * Helper for observing [BrowserStore] and show an in-app crash reporter for tabs with content crashes.
+ *
+ * @param browserStore [BrowserStore] observed for any changes related to [EngineState.crashed].
+ * @param appStore [AppStore] that tracks all content crashes in the current app session until the user
+ * decides to either send or dismiss all crash reports.
+ * @param toolbar [BrowserToolbar] that will be expanded when showing the in-app crash reporter.
+ * @param isToolbarPlacedAtTop [Boolean] based allowing the in-app crash reporter to be shown as
+ * immediately below or above the toolbar.
+ * @param crashReporterView [CrashContentView] which will be shown if the current tab is marked as crashed.
+ * @param components [Components] allowing interactions with other app features.
+ * @param settings [Settings] allowing to check whether crash reporting is enabled or not.
+ * @param sessionId [String] Id of the tab or custom tab which should be observed for [EngineState.crashed]
+ * depending on which [crashReporterView] will be shown or hidden.
+ */
+class CrashContentIntegration(
+    browserStore: BrowserStore,
+    private val appStore: AppStore,
+    private val toolbar: BrowserToolbar,
+    private val isToolbarPlacedAtTop: Boolean,
+    private val crashReporterView: CrashContentView,
+    private val components: Components,
+    private val settings: Settings,
+    private val sessionId: String?
+) : AbstractBinding<BrowserState>(browserStore) {
+    override suspend fun onState(flow: Flow<BrowserState>) {
+        flow.mapNotNull { state -> state.findTabOrCustomTabOrSelectedTab(sessionId) }
```

Comment:
If `sessionId` is null here, should we continue?

Reply:
Yes, I want to continue here based on the following scenario:
- `sessionId` passed to this components is `null` <=> there is no set `customTabSessionId` in BaseBrowserFragment
so the user is in normal browsing => we have to default on checking the `crashed` status of the selected tab.
- `sessionId` passed to this components is not `null` => have to check the state for a particular custom tab - a custom tab cannot be "selected" so just checking for selected would not show this crash reporter for a custom tab.
- when a tab crashes AC will set the `crashed` status for a particular tab - custom or normal.
- user can only see the in-app crash reporter after it selects that tab

So if the user is on a custom tab we'll listen for it's updates otherwise we'll listen for the updates of the currently selected tab after the user selects that tab - and accesses it - and so we'll update (show/hide) the in-app crash reporter only for the custom tab or the currently selected tab.